esr/fitting/test_all_Fisher.py

```diff
@@ -117,24 +117,56 @@ def fop(x):
         start = int(i * max_param - (i - 1) * i / 2)
         deriv[start:start+nparam-i] = Hmat[i,i:]
     
-    if (np.sum(Fisher_diag <= 0.) > 0.) or (np.sum(np.isnan(Fisher_diag)) > 0) or (np.sum(np.isinf(Fisher_diag)) > 0) or (np.sum(Nsteps<1) > 0):
-        for d2, meth in itertools.product(d_list, method_list):
-            if use_relative_dx:
-                Hfun = nd.Hessian(fop, step=np.abs(d2*theta_ML)+1.e-15, method=meth)
-            else:
-                Hfun = nd.Hessian(fop, step = d2, method=meth)
-            Hmat = Hfun(theta_ML)
-            Fisher_diag_tmp = np.array([Hmat[i,i] for i in range(nparam)])
-            Delta_tmp = np.sqrt(12./Fisher_diag_tmp)
-            Nsteps_tmp = abs(np.array(theta_ML))/Delta_tmp
-
-            if (np.sum(Fisher_diag_tmp <= 0.) <= 0) and (np.sum(np.isnan(Fisher_diag_tmp)) <= 0) and (np.sum(np.isinf(Fisher_diag)) <= 0) and (np.sum(Nsteps_tmp<1)<=0):
-                print("\tSucceeded at rectifying Fisher:", fcn_i, d2, meth, Fisher_diag_tmp, theta_ML, Delta_tmp, Nsteps_tmp, flush=True)
-                Fisher_diag = Fisher_diag_tmp
-                Delta, Nsteps = Delta_tmp, Nsteps_tmp
-                start = int(i * max_param - (i - 1) * i / 2)
-                deriv[start:start+nparam-i] = Hmat[i,i:]
-                break
+    n_iter = len(d_list)*len(method_list)
+    if (np.sum(Fisher_diag <= 0.) > 0.) or (np.sum(np.isnan(Fisher_diag)) > 0) or (np.sum(np.isinf(Fisher_diag)) > 0):#  or (np.sum(Nsteps<1) > 0):
+       Fisher_array = np.empty((n_iter, nparam))
+       Hmat_array = np.empty((n_iter, nparam, nparam))
+       e = 0
+       for d2, meth in itertools.product(d_list, method_list):
+           if use_relative_dx:
+               Hfun = nd.Hessian(fop, step = np.abs(d2*theta_ML)+1.e-15, method=meth)
+           else:
+               Hfun = nd.Hessian(fop, step = d2, method=meth)
+           Hmat = Hfun(theta_ML)
+           Hmat_array[e] = Hmat
+           e += 1
+           
+       Hmat_array_f = [] # filter array
+       for matrix in Hmat_array:
+           if not np.any(np.isnan(matrix)) and not np.any(np.isinf(matrix)) and np.all(matrix > 0):
```

Comment:
Why do you have the check `np.all(matrix > 0)`? This is not necessary for a valid Hessian at a maximum. For example, consider a standard multivariate Gaussian. The second derivative of -log(L) is the inverse covariance. This can have negative off-diagonal elements, e.g. ((1, -1/2), (-1/2, 1)) is perfectly valid (since it has eigenvalues of 1/2 and 3/2) but it would fail np.all(matrix > 0). Or worse, if my covariance was already diagonal, the zero off-diagonal entries would cause an error here. So I think this check should be removed.

Reply:
I think so, as this is the same check as the start of the if statement ` if (np.sum(Fisher_diag <= 0.) > 0.)`